### PR TITLE
Feature/equality checking for boxes

### LINKED ISSUE #897: Add semantic equality checks for `CircBox`
And other box types where it's easy to do so.

### PR DESCRIPTION
**Pull request by yao-cqc**
resolves #897

### CODE REVIEW

File: tket/src/Circuit/ToffoliBox.cpp [hunk 1]
```diff
@@ -355,6 +355,37 @@ void ToffoliBox::generate_circuit() const {
       permute(perm, n_, get_op_ptr(rotation_axis_, 1)));
 }
 
+static bool perm_compare(const state_perm_t &a, const state_perm_t &b) {
+  auto ita = a.begin();
+  auto itb = b.begin();
+  for (; ita != a.end(); ita++) {
```
Comment (cqc-alec) [inline tket/src/Circuit/ToffoliBox.cpp:361]: Could refactor this into a static helper function that's called twice with swapped arguments. Up to you.

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Thanks, just a few comments.

File: tket/src/Circuit/Boxes.cpp [hunk 2]
```diff
@@ -205,6 +229,14 @@ Op_ptr ExpBox::transpose() const {
   return std::make_shared<ExpBox>(A_.transpose(), t_);
 }
 
+bool ExpBox::is_equal(const Op &op_other) const {
+  const ExpBox &other = dynamic_cast<const ExpBox &>(op_other);
+  if (id_ == other.get_id()) return true;
+  std::optional<Eigen::MatrixXcd> m = get_box_unitary();
+  std::optional<Eigen::MatrixXcd> other_m = other.get_box_unitary();
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Circuit/Boxes.cpp:236]: Do you mean` A_ == other.A_ and t_ == other.t_ (mod 2*pi)`?

[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Circuit/Boxes.cpp:236]: I'm not sure where to apply that `(mod 2*pi)` other than the eigenvalues.
```
A = np.asarray([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])
B = np.asarray([[-1, 0, 0, 0], [0, -1, 0, 0], [0, 0, -1, 0], [0, 0, 0, -1]])
ebox1 = ExpBox(A, 2.3)
ebox2 = ExpBox(B, -2.3)
```
ebox1 and ebox2 have the same unitary but A!=B

Sorry, I might have misunderstood your comment.

[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/src/Circuit/Boxes.cpp:236]: You're right, it isn't that simple. I suppose the criterion is that
```
t_ * A_ - other.t_ * other.A_ == n * 2 * pi * identity
```
for some integer `n`. But I am not sure about this.

[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/src/Circuit/Boxes.cpp:236]: Happy to keep what you have.

File: tket/src/Circuit/Multiplexor.cpp [hunk 3]
```diff
@@ -411,6 +411,12 @@ op_signature_t MultiplexorBox::get_signature() const {
   return qubits;
 }
 
+bool MultiplexorBox::is_equal(const Op &op_other) const {
+  const MultiplexorBox &other = dynamic_cast<const MultiplexorBox &>(op_other);
+  if (id_ == other.get_id()) return true;
+  return op_map_ == other.op_map_;
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Circuit/Multiplexor.cpp:417]: Absolutely right! Thanks

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**